File: JAVA/25_Shortest_Path/BOJ_05_11657_타임머신.java
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.ArrayList;
import java.util.StringTokenizer;

class point {
    int from, to, cost;

    public point(int from, int to, int cost) {
        this.from = from;
        this.to = to;
        this.cost = cost;
    }
}

public class Main {

    final static int INF = Integer.MAX_VALUE;
    static int n, m;
    static long[] min;
    static ArrayList<point> list;

    public static boolean bellmanFord() {
        for (int i = 0; i < n - 1; i++) {
            for (int j = 0; j < m; j++) {
                point bus = list.get(j);
                if (min[bus.from] == INF) continue;
                min[bus.to] = Math.min(min[bus.to], min[bus.from] + bus.cost);
            }
        }
        for (int j = 0; j < m; j++) {
            point bus = list.get(j);
            if (min[bus.from] != INF && min[bus.to] > min[bus.from] + bus.cost) {
                return true;
            }
        }
        return false;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        list = new ArrayList<>();
        min = new long[n + 1];
        Arrays.fill(min, INF);
        min[1] = 0;
        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine(), " ");
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            int c = Integer.parseInt(st.nextToken());
            list.add(new point(a, b, c));
        }
        if (bellmanFord()) {
            System.out.println(-1);
        } else {
            StringBuilder sb = new StringBuilder();
            for (int i = 2; i <= n; i++) {
                sb.append(min[i] != INF ? min[i] : -1).append("\n");
            }
            System.out.print(sb);
        }
    }
}

File: C/25_Shortest_Path/BOJ_05_11657_타임머신.c
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#define math_min(a, b) a < b ? a : b
typedef struct {
	int from, to, cost;
} point;
point* list;
int n, m;
int INF = INT_MAX;
long long int* min;
int bellmanFord() {
	for (int i = 0; i < n - 1; i++) {
		for (int j = 0; j < m; j++) {
			point bus = list[j];
			if (min[bus.from] == INF) continue;
			min[bus.to] = math_min(min[bus.to], min[bus.from] + bus.cost);
		}
	}
	for (int j = 0; j < m; j++) {
		point bus = list[j];
		if (min[bus.from] != INF && min[bus.to] > min[bus.from] + bus.cost) {
			return 1;
		}
	}
	return 0;
}
main() {
	scanf("%d %d", &n, &m);
	min = (long long int*)malloc(sizeof(long long int) * (n + 1));
	for (int i = 2; i <= n; i++) {
		min[i] = INF;
	}
	min[1] = 0;
	list = (point*)malloc(sizeof(point) * m);
	for (int i = 0; i < m; i++) {
		scanf("%d %d %d", &list[i].from, &list[i].to, &list[i].cost);
	}
	if (bellmanFord()) {
		printf("-1");
	}
	else {
		for (int i = 2; i <= n; i++) {
			printf("%d\n", min[i] != INF ? min[i] : -1);
		}
	}
}